File: category-service/routes/category.go
package routes

import (
	"encoding/json"
	"log"
	"net/http"

	"category-service/models"

	"github.com/gorilla/mux"
	"gorm.io/gorm"
)

type Router struct {
	DB *gorm.DB
}

func NewRouter(db *gorm.DB) *mux.Router {
	router := mux.NewRouter()
	handler := &Router{DB: db}

	router.HandleFunc("/api/category", handler.GetCategoriesHandler).Methods("GET")
	router.HandleFunc("/api/category", handler.CreateCategoryHandler).Methods("POST")

	return router
}

func (r *Router) GetCategoriesHandler(w http.ResponseWriter, req *http.Request) {
	var categories []models.Category

	// Use GORM to find all categories
	result := r.DB.Find(&categories)
	if result.Error != nil {
		http.Error(w, "Failed to query categories", http.StatusInternalServerError)
		log.Printf("Database query error: %v", result.Error)
		return
	}

	// Set response headers and encode JSON
	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(categories); err != nil {
		http.Error(w, "Failed to encode response as JSON", http.StatusInternalServerError)
		log.Printf("JSON encoding error: %v", err)
	}
}

func (r *Router) CreateCategoryHandler(w http.ResponseWriter, req *http.Request) {
	var category models.Category

	// Decode the incoming JSON request
	if err := json.NewDecoder(req.Body).Decode(&category); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		log.Printf("JSON decoding error: %v", err)
		return
	}

	// Create the category in the database
	result := r.DB.Create(&category)
	if result.Error != nil {
		http.Error(w, "Failed to create category", http.StatusInternalServerError)
		log.Printf("Database create error: %v", result.Error)
		return
	}

	// Set response headers and encode the created category
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	if err := json.NewEncoder(w).Encode(category); err != nil {
		http.Error(w, "Failed to encode response as JSON", http.StatusInternalServerError)
		log.Printf("JSON encoding error: %v", err)
	}
}

File: src/handlers/category.ts
import { Router } from "express";
import { TransactionV2Service } from "../services/TransactionV2Service";
import { Category } from "../types/category";

export function mountCategoryRoutes(router: Router) {
    router.get("/monthly-expenses-by-category", async (req, res) => {
      try {
        let result: Record<string, number>
        const service = new TransactionV2Service()
        result = await service.getMonthlyExpensesByCategory()
    
        res.json(result)
      } catch (error) {
        console.error(error)
        res.status(500).json({ error: "Failed to fetch data /total-values-by-category", cause: error })
      }
    })


    // Category Service Proxy Routes
    router.get("/category", async (req, res) => {
      try {
        const transactionV2Service = new TransactionV2Service()
        const response = await transactionV2Service.get("/categories")
        res.status(response.status).json(response.data)
      } catch (error: any) {
        console.error(error)
        res.status(error?.status || 500).json({
          error: "Failed to proxy request to GET /category",
          cause: error?.response?.data ?? error,
        })
      }
    })
    
    router.post("/category", async (req, res) => {
      try {
        const transactionV2Service = new TransactionV2Service()
        const response = await transactionV2Service.post("/categories", req.body)
        res.status(response.status).json(response.data)
      } catch (error: any) {
        console.error(error)
        res.status(error?.status || 500).json({
          error: "Failed to proxy request to POST /category",
          cause: error?.response?.data ?? error,
        })
      }
    })
}